feature: update repo

mrnkr opened the pull request:
Hello again!

As @bart-degreed suggested in #2 , this PR is the first one of at least three that will come. This one contains only the update to the latest version of JADNC and the new implementation of `MongoEntityRepository` that comes with it.

For reference, I quote:

> Update to the latest version of JADNC with .NET Core 3.1. This would include your MongoDB repository and basic getting-started instructions, but without the example/getting-started/test projects. Sort of like: we just updated the codebase to make it run, no guarantees about what actually works.

Let me know if I need to polish anything before moving on to the second PR, I'll try to have it all ready ASAP :)

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:
```diff
@@ -0,0 +1,182 @@
+            _resourceFactory = resourceFactory;
+        }
+
+        private IMongoCollection<TResource> Collection => _db.GetCollection<TResource>(typeof(TResource).Name);
+        private IMongoQueryable<TResource> Entities => Collection.AsQueryable();
```
Instead of this, a protected virtual method `GetAll()` should be provided (and used from all places), so that users can override and put custom logic in there. For example, to add an extra filter condition for authorization.

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:
```diff
@@ -0,0 +1,182 @@
+            return resources.FirstOrDefault();
+        }
+
+        public virtual async Task UpdateAsync(TResource requestResource, TResource databaseResource, CancellationToken cancellationToken)
```
Please adapt parameter names to match with interface signature:
```
public virtual async Task UpdateAsync(TResource resourceFromRequest, TResource resourceFromDatabase, CancellationToken cancellationToken)
```

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:
```diff
@@ -0,0 +1,182 @@
+            throw new NotImplementedException();
+        }
+
+        protected virtual IMongoQueryable<TResource> ApplyQueryLayer(QueryLayer layer)
```
Parameter null check is missing here, as well as in other public methods.
Also please move this method below CountAsync(), where it is being used for the first time. This keeps it more in sync with the main repo too.

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoEntityRepository.cs:
```diff
@@ -0,0 +1,182 @@
+            throw new NotImplementedException();
+        }
+
+        public object this[string name] => throw new NotImplementedException();
```
Please move this indexer property to above the constructor.

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/Extensions/MongoIQueryableExtensions.cs:
```diff
@@ -1,37 +1,20 @@
-        {
-            return await IAsyncCursorSourceExtensions.SingleOrDefaultAsync(ToMongoQueryable(queryable));
-        }
+        public static async Task<IReadOnlyList<T>> ToListAsync<T>(this IQueryable<T> queryable) =>
```
There is only a single usage of this method, so I think would be better to move it into the repository itself as a private method. That makes the code simpler, because the check for `IMongoQueryable` can be removed. And it no longer hides the fact that `CancellationToken` is not being passed, which it should take.

bart-degreed reviewed (changes requested):
See the various comments.

bart-degreed commented on README.md:
```diff
@@ -0,0 +1,76 @@
+            return client.GetDatabase(Configuration.GetSection("DatabaseSettings:Database").Value);
+        });
+
+        services.AddScoped<IResourceRepository<Book, string>, MongoEntityRepository<Book, string>>();
```
Also, this:
```c#
services.AddScoped<IResourceRepository<Book, string>, MongoEntityRepository<Book, string>>();
```
does not work with JADNC v4-final. It needs to be registered on `IResourceRepository<>`, `IResourceReadRepository<>` and `IResourceWriteRepository<>`. We have a convenient helper extension method for that:
```c#
services.AddResourceRepository<MongoEntityRepository<Book, string>>();
```

bart-degreed commented on README.md:
```diff
@@ -0,0 +1,76 @@
+            return client.GetDatabase(Configuration.GetSection("DatabaseSettings:Database").Value);
+        });
+
+        services.AddScoped<IResourceRepository<Book, string>, MongoEntityRepository<Book, string>>();
```
On second thought, MongoRepository or MongoDbRepository may be more appropriate.

mrnkr commented on src/JsonApiDotNetCore.MongoDb/JsonApiDotNetCore.MongoDb.csproj:
```diff
@@ -1,12 +1,29 @@
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <VersionPrefix>4.0.0-rc</VersionPrefix>
```
Yes, it was me who published the package. I did it because the PR had been stale for a while but I'll be more than happy to transfer ownership of it to you.

bart-degreed replied:
Thanks, I would like to accept ownership of it. I can look into setting up cibuild somewhere next week. We've been having lots of timeouts with Travis CI recently, so I would prefer to use AppVeyor which is always fast. Also that enables me to reuse most of the existing scripts.

mrnkr commented:
I think I addressed all the comments. Let me know if there is anything else for me to do!

By the way @bart-degreed , you have no idea how much I've learnt from your comments in both this repo and JsonApiDotNetCore so thanks a lot for that!

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoDbRepository.cs:
```diff
@@ -0,0 +1,230 @@
+    }
+    
+    /// <summary>
+    /// Implements the foundational repository implementation that uses MongoDB.
```
It looks like you took this comment from the main project, but it is wrong there. They should be the same on both types:
```
Implements the foundational Repository layer in the JsonApiDotNetCore architecture that uses MongoDB.
```

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoDbRepository.cs:
```diff
@@ -0,0 +1,230 @@
+        public virtual async Task<IReadOnlyCollection<TResource>> GetAsync(QueryLayer layer,
+            CancellationToken cancellationToken)
+        {
+            if (layer == null) throw new ArgumentNullException(nameof(layer));
+            var list = await ApplyQueryLayer(layer).ToListAsync(cancellationToken);
+            return list.AsReadOnly();
```
Please add a blank line after input validation. Also a better name for variable `list` (the type) would be `resources` (what it represents).

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoDbRepository.cs:
```diff
@@ -0,0 +1,230 @@
+
+            if (!result.IsAcknowledged)
+            {
+                throw new DataStoreUpdateException(new Exception("Delete operation was not acknowledged by MongoDB"));
```
Sentences should end with a dot. And it would be nice to include the `id` value here, too. Suggestion:
```
throw new DataStoreUpdateException(new Exception($"Failed to delete document with id '{id}', because the operation was not acknowledged by MongoDB."));
```

bart-degreed commented on src/JsonApiDotNetCore.MongoDb/MongoDbRepository.cs:
```diff
@@ -0,0 +1,230 @@
+
+            if (result.DeletedCount == 0)
+            {
+                throw new DataStoreUpdateException(new Exception($"No documents were deleted. The id that was being sought was: {id}"));
```
Suggestion:
```
throw new DataStoreUpdateException(new Exception($"Failed to delete document with id '{id}', because it does not exist."));
```

bart-degreed reviewed (changes requested):
> By the way @bart-degreed , you have no idea how much I've learnt from your comments in both this repo and JsonApiDotNetCore so thanks a lot for that!

Thank you, that's really great to hear! Got a few more remarks, would be nice if you can take a look at them.

mrnkr commented:
@bart-degreed I've just added you as an owner of the package on NuGet and did the changes you requested in your last review.
